### PR TITLE
feat: Asynchronous Streaming for Source and Sink in Numaflow Core

### PR DESCRIPTION
**Pull request by yhl25**

### CODE REVIEW

File: rust/numaflow-core/src/message.rs [hunk 1]
```diff
@@ -103,7 +79,7 @@ impl TryFrom<async_nats::Message> for Message {
         let event_time = Utc::now();
         let offset = None;
         let id = MessageID {
-            vertex_name: get_vertex_name().to_string(),
+            vertex_name: config::get_vertex_name().to_string(),
```
Comment (whynowy) [inline rust/numaflow-core/src/message.rs:82]: Having `get_vertex_name()` in this `try_from` function is weird. Do we have to implement `try_from`, or a until function to convert Message (with vertexName in the signature) also works?

File: rust/numaflow-core/src/message.rs [hunk 2]
```diff
@@ -319,7 +303,7 @@ impl TryFrom<read_response::Result> for Message {
             offset: Some(source_offset.clone()),
             event_time: utc_from_timestamp(result.event_time),
             id: MessageID {
-                vertex_name: get_vertex_name().to_string(),
+                vertex_name: config::get_vertex_name().to_string(),
```
Comment (whynowy) [inline rust/numaflow-core/src/message.rs:306]: Same question.

File: rust/numaflow-core/src/config/monovertex.rs [hunk 3]
```diff
@@ -94,6 +94,7 @@ impl MonovertexConfig {
             .as_ref()
             .and_then(|source| source.transformer.as_ref())
             .map(|_| TransformerConfig {
+                concurrency: batch_size as usize, // FIXME: introduce a new config called udf concurrency in the spec
```
Comment (whynowy) [inline rust/numaflow-core/src/config/monovertex.rs:97]: Do we need to?

File: rust/numaflow-core/src/message.rs (hunk 1 above)
Comment (vigith) [reply rust/numaflow-core/src/message.rs:82]: what do you mean by an `until` function?

File: rust/numaflow-core/src/config/monovertex.rs (hunk 3 above)
Comment (vigith) [reply rust/numaflow-core/src/config/monovertex.rs:97]: this is to make sure we can achieve X workers in parallel immaterial of the messages from previous vertex. we do read-ahead to achieve X concurrency/parallelism.

**Review by vigith — APPROVED** [review]
please remove self.clone() in source.rs

**Pull request merged**